src/tests/JIT/Regression/JitBlue/Runtime_34937/Runtime_34937.cs

```diff
@@ -4,55 +4,87 @@
 using System;
 using System.Runtime.CompilerServices;
 
-class Program
+class Program<T>
 {
     [MethodImpl(MethodImplOptions.NoInlining)]
-    static uint PerformMod_1(uint i)
+    public static uint PerformMod_1(uint i)
     {
+        // X64-FULL-LINE:      mov [[REG0:[a-z]+]], [[REG1:[a-z0-9]+]]
+        // X64-FULL-LINE-NEXT: and [[REG0]], 7
+
+        // ARM64-FULL-LINE: and w0, w0, #7
+
         return i % 8;
     }
 
     [MethodImpl(MethodImplOptions.NoInlining)]
-    static int PerformMod_2(int i)
+    public static int PerformMod_2(int i)
     {
+        // X64-FULL-LINE:              mov [[REG0:[a-z]+]], [[REG1:[a-z]+]]
+        // X64-FULL-LINE-NEXT:         sar [[REG0]], 31
+        // X64-FULL-LINE-NEXT:         and [[REG0]], 15
+        // X64-FULL-LINE-NEXT:         add [[REG0]], [[REG1]]
+        // X64-FULL-LINE-NEXT:         and [[REG0]], -16
+        // X64-WINDOWS-FULL-LINE-NEXT: mov [[REG2:[a-z]+]], [[REG1]]
```

Comment:
We should ask. 

@jakobbotsch @kunalspathak - I wrote a test here that showed a case where the codegen is different in windows in that it adds an extra `mov` instruction. We don't know if this is necessary or not.

Reply:
It's not clear to me, seems like maybe there is some suboptimal preferencing in LSRA? The argument is in a different register on these two ABIs so seems like that somehow ends up affecting it.

Reply:
During register assignment, if we have more than 1 register available to assign, we try to narrow them down using various heuristics until we are left with just 1 register that we assign. However, if there is a tie-breaker, we just pick a register in the register order. 

At the problematic site, both the registers were assigned using `REG_ORDER` meaning we found more than 1 register to be assigned and we decided based on the ordering of register.

<img width="1539" alt="image" src="https://user-images.githubusercontent.com/12488060/192043916-8aa09103-ae22-49d3-a113-e9e0dbf2d5c5.png">


It so happened that for Unix/x64, `EDI` (which was the original register) comes first, but for windows, `EDX` comes before `ECX` and we prefer it. 

https://github.com/dotnet/runtime/blob/d0559e6e33906e4e6db675e646b47ab36eb5bffc/src/coreclr/jit/targetamd64.h#L191-L203

We can probably do better (add a new heuristic probably) by reutilizing existing assigned register if they won't be used any further, instead of assigning a new register.

Reply:
Ok, that is starting to make a lot more sense. Thank you for the detailed information @kunalspathak .

Reply:
@kunalspathak Are those heuristics mostly/all correctness based (as in, don't reuse a register that already has a value), or is it possible that one of the heuristics should have helped here but didn't for some reason?

Another possible heuristic would be trying to use `eax` in order to feed the `ret` without a `mov`.  However, that would require -not- using it above, so maybe it already exists but couldn't be used.  In this case, it's not as good because it would lose the short encodings for `and`.

Reply:
> @kunalspathak Are those heuristics mostly/all correctness based (as in, don't reuse a register that already has a value), or is it possible that one of the heuristics should have helped here but didn't for some reason?

Well, I would say they are for performant reasons. If you end of reusing same sets of registers that already hold a live variable you end up spilling the live value to the stack (not good) or if we reuse the registers from inactive interval, we end up copying/reloading (again, not good).